### PR TITLE
Media: Add new screen

### PR DESCRIPTION
**Pull request by kean**
Part of https://github.com/wordpress-mobile/WordPress-iOS/issues/21457

This PR introduces the first iteration of the new `MediaViewController` screen that will eventually replace the WPMediaPicker-based implementation. The screen doesn't do much yet: it just displays the assets already added to the Core Data database and loads their thumbnails. This PR focuses heavily on loading the thumbnails. 

The refreshed system for loadings thumbnails (`MediaImageService`) uses the existing pieces from (soft-deprecated) `MediaThumbnailService`, but adds async/await on top of it in order to support cooperative cancellation. There are a few major performance fixes/improvements to the image loading:

- Adds prefetching powered by `UICollectionViewDataSourcePrefetching` (addresses https://github.com/wordpress-mobile/WordPress-iOS/issues/4288)
- Fix main thread [performance issues](https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/7180d9e4-a6c1-4a34-a6e3-969ba8f166e8) by moving image decompression to the background
- Move disk IO in availableThumbnail to the background
- Removes unnecessary `coreDataStack.mainContext.perform` hops and use `@MainActor` to enforce thread confinement 
- Removes unnecessary reloads on `.update` changes
- Add cancellation so it you scroll quickly you won’t be bogged down by all the outsanding requests (addresses https://github.com/wordpress-mobile/WordPress-iOS/issues/14528)
- Maybe more I forgot about

Fixes:

- Fix missing placeholder when image downloading. If you scroll to the empty list, it appears empty.

Notes: 

- I tried diff able data source, but it was too slow for incremental updates in large libraries (I tested on Jetpack Mobile)

There are no changes to the production code – only additions – so it should be safe to merge and continue iterating on it in the future PRs.

## To test:

It's an early prototype, so testing anything would likely be a waste of time. Let's do it in separate PRs once at least some of the main pieces are fully developed.

If you want to see the new screen, you'll need to show it programmatically:

```swift
DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(1)) {
    let vc = MediaViewController()
    let nav = UINavigationController(rootViewController: vc)
    self.present(nav, animated: true)
}
```

## Regression Notes
1. Potential unintended areas of impact: n/a
2. What I did to test those areas of impact (or what existing automated tests I relied on): manual tests
3. What automated tests I added (or what prevented me from doing so): n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

File: WordPress/Classes/Services/LegacyMediaThumbnailService.swift [hunk 1]
```diff
@@ -0,0 +1,216 @@
+import Foundation
+
+/// A service for handling the process of retrieving and generating thumbnail images
+/// for existing Media objects, whether remote or locally available.
+///
+final class LegacyMediaThumbnailService: NSObject {
```
Comment (kean) [inline WordPress/Classes/Services/LegacyMediaThumbnailService.swift:6]: I'm not sure why it's showing it as a nice file. I simply renamed it.

File: WordPress/Classes/Services/MediaImageService.swift [hunk 2]
```diff
@@ -0,0 +1,196 @@
+import Foundation
+
+/// A service for handling the process of retrieving and generating thumbnail images
+/// for existing Media objects, whether remote or locally available.
+final class MediaImageService: NSObject {
```
Comment (kean) [inline WordPress/Classes/Services/MediaImageService.swift:5]: This class is a replacement for [MediaThumbnailService.swift](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/Services/MediaThumbnailService.swift), which I haven't touched to make sure it doesn't introduce any regressions. 

The previous implementation was pretty hard to follow because of all of the closures, so it was a perfect use-case for async/await. But the main reason I added it was to support cancellation to address https://github.com/wordpress-mobile/WordPress-iOS/issues/14528.

It also takes advantage of `Task.detached` for some background tasks. The great thing about the new concurrency system is that it automatically limits the number of threads.

File: WordPress/Classes/ViewRelated/Media/Library/MediaCollectionCellViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+@MainActor
+final class MediaCollectionCellViewModel {
+    var onLoadingFinished: ((UIImage?) -> Void)?
+    let mediaID: TaggedManagedObjectID<Media>
+    var mediaType: MediaType
+
+    private let media: Media
+    private let service: MediaImageService
+    private let cache: MemoryCache
```
Comment (kean) [inline WordPress/Classes/ViewRelated/Media/Library/MediaCollectionCellViewModel.swift:11]: I'll move this cache to the new service in one of the future PRs.

File: WordPress/Classes/Services/MediaImageService.swift [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        }
+
+        let imageURL = try await imageURL(for: media, preferredSize: preferredSize)
+        return try await Task.detached {
+            guard let image = UIImage(data: try Data(contentsOf: imageURL)) else {
+                throw URLError(.cannotDecodeContentData)
+            }
+            // Forces decompression (or bitmapping) to happen in the background.
+            // It's very expensive for some image formats, such as JPEG.
+            return image.preparingForDisplay() ?? image
```
Comment (momo-ozawa) [inline WordPress/Classes/Services/MediaImageService.swift:38]: [Q] Is Task.detached {} what forces decompression to happen in the background?

**Review by momo-ozawa — APPROVED** [review]
Learning lots from reviewing your PR! Trying to get up to speed with modern concurrency :P

The media cells are missing some tags (i.e. video duration, gif) - but I assume you're handling that in a separate PR. 

Current | New 
-- | --
 |

File: WordPress/Classes/Services/MediaImageService.swift (hunk 4 above)
Comment (kean) [reply WordPress/Classes/Services/MediaImageService.swift:38]: `Task.detached` sends the work to one of the background threads. It uses the new concurrency system, so you never risk running into a thread explosion problem, unlike `DispatchQueue.global`.

[`prepareForDisplay`](https://developer.apple.com/documentation/uikit/uiimage/3750834-preparingfordisplay) is a relatively new API (iOS 15) that decompressed the images.

**Comment by kean** [comment]
> The media cells are missing some tags (i.e. video duration, gif) - but I assume you're handling that in a separate PR.

Yeah, there is a lot of work left to do.
